### PR TITLE
Add Alpine and musl support

### PR DESCRIPTION
**Pull request by giulio-coa**
## Thank you for your contribution to the Pi-hole Community! 

Please read the comments below to help us consider your Pull Request.

We are all volunteers and completing the process outlined will help us review your commits quicker.

**Please make sure you**
 1. Base your code and PRs against the repositories developmental branch.
 2. [Sign Off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits as we enforce the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions

---
- **What does this PR aim to accomplish?:**

I would like to be able to add support for `musl` and Alpine to the various architectures present

- **How does this PR accomplish the above?:**

I have added support for almost all architectures
Unfortunately there are no alpine images that support `armv4` or `32bit armv8`

- **What documentation changes (if any) are needed to support this PR?:**

*Replace this with a detailed list of any necessary changes*

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
- [x] I have read the above and my PR is ready for review. _Check this box to confirm_

### CODE REVIEW

**Comment by giulio-coa** [comment]
Sorry
I changed the branch name but it is the same of #49

**Comment by PromoFaux** [comment]
>>Can you rebase on master please and we'll see what's what

>Sorry, but I didn't follow your reasoning 😅

Builds are failing due to not being able to log into Docker Hub.  secrets are protected on PR's from forks. That said, PR's don't get pushed to docker hub in any case so there is no need for the login step on a PR

image

I got the boolean logic the wrong way around initially but that should now be fixed: please rebase again on master, so that your PR contains this commit: https://github.com/pi-hole/docker-base-images/commit/9fde59e7b2b0be56cd2b16c35d4bebccd0df3678

File: ftl-build/x86_64-musl/Dockerfile [hunk 1]
```diff
@@ -1,5 +1,5 @@
-ARG ALPINE_VER=3.14
-FROM alpine:${ALPINE_VER}
+ARG ALPINE_VER
+FROM amd64/alpine:${ALPINE_VER:-v3.14}
```
Comment (PromoFaux) [inline ftl-build/x86_64-musl/Dockerfile:2]: Why change this from `VAR=value` to variable expansion?

Comment (giulio-coa) [reply ftl-build/x86_64-musl/Dockerfile]: I remembered seeing on the Docker official site that it was better to use that syntax, but I was wrong
It isn't said anywhere

**Review by PromoFaux — CHANGES_REQUESTED** [review]
Can we focus on one thing at a time, please? Please remove any changes to the docker base image stuff (including workflow files) and keep only the changes for ftl-build.

I see what you are trying to do with the docker base image stuff but to be honest it looks a bit hacky - there is probably a better way to do this without duplicating sections of the yaml and tweaking it to read debian-based/alpine-based. That's academic, for now - we can come back to it at a later stage

The builds are going to fail, too, because PRs don't have access to the secrets. Can you please set a `DOCKERHUB_NAMESPACE` secret on your fork - maybe that will prevent the issue, though I doubt it. Perhaps I need to set up this repo to be more friendly to external PRs (though this is the first we've ever had!)

**Comment by PromoFaux** [comment]
@giulio-coa you also need to test you changes before submitting a PR. I just cherry picked your commits for only the FTL stuff, and none of the images are building. all musl variants have :

```
Error: buildx failed with: error: failed to solve: arm64v8/alpine:v3.14: docker.io/arm64v8/alpine:v3.14: not found
```

The non-musl ones are not building either, but I don't think that is something you have introduced because you haven't touched those

**Comment by giulio-coa** [comment]
> @giulio-coa you also need to test you changes before submitting a PR. I just cherry picked your commits for only the FTL stuff, and none of the images are building. all musl variants have :
> 
> ```
> Error: buildx failed with: error: failed to solve: arm64v8/alpine:v3.14: docker.io/arm64v8/alpine:v3.14: not found
> ```
> 
> The non-musl ones are not building either, but I don't think that is something you have introduced because you haven't touched those

I foolishly tested with `3.14` instead of `v3.14`

**Comment by PromoFaux** [comment]
Happens to the best of us. I'm just making some other changes locally to fix the other build issues. (`pdns-backend-sqlite3_4.4.1` has gone missing from powerdns' servers. Likewise we have accidentally housekept away a few libraries from ftl.pi-hole.net/libraries - so I need to hunt down the missing files)

Once I've done that I'll continue in #50

**Comment by PromoFaux** [comment]
OK, `master` is now fixed and the existing images should be building again. One thing to note is that pdns is only needed on the x86_32/64 images - as full test suite is not run on arm containers. I'll sort that out

**Comment by PromoFaux** [comment]
That all said - there is going to be a lot more work involved in all this than what you have (so far) done. By the looks of things you just have copied the x86_64-musl Dockerfile to the other arch's and hoped for the best :)

Examine the non-alpine dockerfiles in detail - and note how they differ depending on the arch that FTL is being cross-compiled for. Make the changes locally. Attempt to build each individual arch one by one locally until you have a working process, and only then push back up to Github for further review. There is little sense in pushing work to Github that we know is going to fail the CI

Not trying to put you off.. just, this needs a lot more effort than I  personally have time to give it

**Comment by giulio-coa** [comment]
> The builds are going to fail, too, because PRs don't have access to the secrets. Can you please set a `DOCKERHUB_NAMESPACE` secret on your fork - maybe that will prevent the issue, though I doubt it. Perhaps I need to set up this repo to be more friendly to external PRs (though this is the first we've ever had!)

How can I do it ?

**Comment by giulio-coa** [comment]
> The builds are going to fail, too, because PRs don't have access to the secrets. Can you please set a `DOCKERHUB_NAMESPACE` secret on your fork - maybe that will prevent the issue, though I doubt it. Perhaps I need to set up this repo to be more friendly to external PRs (though this is the first we've ever had!)

I guess I have set the secret wrong
How did you set it up ?

**Comment by PromoFaux** [comment]
Should be somewhere like : https://github.com/goulio-coa/docker-base-images/settings/secrets/actions/new

You'll need to set up a variable called `DOCKERHUB_NAMESPACE` and give it a value to match your namespace on docker hub so that anything you build on your fork will be pushed to your own namespace.

However, like I said, that's going to be the least of your worries at the moment - your big challenge is going to be in getting each of the arches to _actually build_ FTL correctly

Going to close this out for now as it's not really in a state to be reviewed

**Pull request closed without merging**